keps/sig-api-machinery/2896-openapi-v3/README.md

```diff
@@ -0,0 +1,563 @@
+
+Additionally, since kubernetes aggregate the OpenAPI v2 spec for all specs into one endpoint, whenever a spec is modified the entire spec is updated and all clients will need to redownload the spec. Depending on the size and number of CRDs, this is not an inexpensive operation.
+
+### Goals
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Support publishing and aggregating OpenAPI v3 for all kubernetes types
+- Published v3 spec should be a lossless representation of the data and no fields should be stripped for both built-in types and CRDs
+- Instead of serving the entire OpenAPI spec at one endpoint, separate the spec by group and only serve the resources required by each group
+
+#### Transparency in the OpenAPI
+
+There are a couple of important fields part of the CRD structural schema that are stripped when we publish v2: `oneOf`, `anyOf`, `nullable`, `default`. We would like to keep all fields and publish v3 specs without any information loss for both built-in types and CRDs.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Accommodating new OpenAPI v3 fields outside the Schema object that do not affect validity of the spec. These are nice to have fields that can be added later.
+- Consumers of the OpenAPI (eg: Server Side Apply, client-go, kubectl, etc) will eventually need to be updated to consume v3. This is outside the scope of this KEP.
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal is to publish OpenAPI v3 schemas at the `/openapi/v3/apis/{group}` endpoint for all resources. Instead of aggregating everything into one schema, we will publish separate schemas per resource group.  The aggregated schema will not be published at the openapi/v3 endpoint and the endpoint will instead be used for discovery purposes. For users who still need to retrieve the entire OpenAPI spec (not recommended), we will offer a client side utility to aggregate all the endpoints.
```

Comment:
as written above: per version would make this much easier on the aggregator side.

Reply:
we could make it `openapi/v3/{group}/{version}`

Reply:
Changing this to group/version